src/test/java/com/extendaretail/vertx/gcp/pubsub/v1/PubSubServiceImplTest.java

```diff
@@ -29,6 +32,49 @@ void testProxyCall(Vertx vertx, VertxTestContext testContext, Tooling tooling) {
     service.publish(message).onComplete(testContext.succeedingThenComplete());
   }
 
+  @Test
+  @Timeout(5000)
+  void testJsonObjectDataShouldTakePrecedenceOverBufferMessage(
+      Vertx vertx, VertxTestContext testContext, Tooling tooling) { // NOSONAR
+    PubSubServiceImpl service = new PubSubServiceImpl(vertx);
+    service.getPublishers().put(tooling.getTopicId(), tooling.getPublisher());
+
+    String jsonPayload = "{\"message\":\"Hello World\"}";
+    Buffer bufferMessageToBeIgnored =
+        Buffer.buffer("ByteArrayMessage".getBytes(StandardCharsets.UTF_8));
+
+    PubSubMessage message =
+        new PubSubMessage()
+            .setTopic(tooling.getTopicId())
+            .setMessage(new JsonObject(jsonPayload))
+            .setBufferMessage(bufferMessageToBeIgnored);
+
+    service.publish(message).onComplete(testContext.succeedingThenComplete());
```

Comment:
`testContext.succeedingThenComplete()` completes the whole test, and will potentially disregard the outcome of the following assert. Use this instead:

```java
Checkpoint publish = testContext.checkpoint();
Checkpoint assertLastMessage = testContext.checkpoint();

service.publish(message).onSuccess(ignore -> publish.flag());

testContext.verify(() -> {
  assertThat(tooling.receiveLastMessage().getMessage().getData().toStringUtf8()).isEqualTo(jsonPayload);
  assertLastMessage.flag();
});
```
All asserts needs to be within a (rather bulky) `testContext.verify` clause for them to be catched, otherwise the testContext is not aware of the potential failures.

We need to "reach" two places, so we must not complete the testcontext, the way to actually complete it is to use checkpoints, when more than one "assertion" needs to be made.

In essence; either use one call to `testContext.succeedingThenComplete()` or use multiple checkpoints.

Reply:
thanks for the explanation 👍 That makes sense